### PR TITLE
remove stale serialized dags

### PR DESCRIPTION
**Pull request by pingzh**
the processor should clean up stale serialized dags after parsing a dag file. The staleness comes from deleting a dag, renaming dag_id in a dag file.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: airflow/models/serialized_dag.py [hunk 1]
```diff
@@ -220,6 +220,36 @@ def remove_dag(cls, dag_id: str, session: Session = None):
...
+
+        :return: a list of inactive dag ids
+        :rtype: list[str]
+        """
+        file_path_hash = DagCode.dag_fileloc_hash(dag_file_path)
+        # there is a db index on fileloc_hash
+        rows = session.query(cls.dag_id).filter(cls.fileloc_hash == file_path_hash).all()
+        dag_ids_from_db = [i[0] for i in rows]
+
+        dag_ids_to_remove = [dag_id for dag_id in dag_ids_from_db if dag_id not in active_dag_ids]
```
Comment (ashb) [inline airflow/models/serialized_dag.py]: It's unlikely, but theoretically possible that we could have a hash collision, so we should also check the full path matches.

File: airflow/models/serialized_dag.py [hunk 2]
```diff
@@ -220,6 +220,36 @@ def remove_dag(cls, dag_id: str, session: Session = None):
...
+
+        if dag_ids_to_remove:
+            # delete one by one to avoid locking the table
```
Comment (ashb) [inline airflow/models/serialized_dag.py]: Does this apply to all DBs or just Mysql?

File: airflow/models/serialized_dag.py (hunk 1 above)
Comment (pingzh) [reply airflow/models/serialized_dag.py]: good call. but i don't think it is necessary to check here. if we think collision is an issue, we might want to change the way it is generated. let me know your thoughts

File: airflow/models/serialized_dag.py (hunk 2 above)
Comment (pingzh) [reply airflow/models/serialized_dag.py]: good question. I am not sure about the other dbs, as we only use mysql. but i think deleting with small batch could have smaller load in db.

File: airflow/models/serialized_dag.py (hunk 1 above)
Comment (ashb) [reply airflow/models/serialized_dag.py]: The index is non-unique as any kind of hashing is going to have _some_ risk of collision.

File: airflow/models/serialized_dag.py [hunk 3]
```diff
@@ -220,6 +220,36 @@ def remove_dag(cls, dag_id: str, session: Session = None):
...
+    ) -> List[str]:
+        """
+        Deletes serialized_dags whose dag_ids are not in the active_dag_ids for a given dag_file_path
+
+        :return: a list of inactive dag ids
+        :rtype: list[str]
+        """
+        file_path_hash = DagCode.dag_fileloc_hash(dag_file_path)
+        # there is a db index on fileloc_hash
+        rows = session.query(cls.dag_id).filter(cls.fileloc_hash == file_path_hash).all()
```
Comment (ashb) [inline airflow/models/serialized_dag.py]:
```suggestion
        rows = session.query(cls.dag_id).filter(cls.fileloc_hash == file_path_hash, cls.fileloc == dag_file_path).all()
```

This should be all it takes -- use the DB index to restrict the rows, and then a non-indexed search to only get the actual rows we want.

File: airflow/models/serialized_dag.py (hunk 1 above)
Comment (pingzh) [reply airflow/models/serialized_dag.py]: after reading the implementation of `dag_fileloc_hash`, it makes sense to do the check. updated.

**Comment by pingzh** [comment]
@ashb could you please take a look when you have chance? thanks

**Comment by potiuk** [comment]
I think that one should wait for 2.4.0dev - adding more to 2.3.0 now unless it's really necessary is a bit risky :D

**Comment by pingzh** [comment]
@potiuk since 2.3 is released. can you please take a look at this PR?

File: airflow/models/serialized_dag.py [hunk 4]
```diff
@@ -220,6 +220,35 @@ def remove_dag(cls, dag_id: str, session: Session = None):
...
+        :rtype: list[str]
+        """
+        file_path_hash = DagCode.dag_fileloc_hash(dag_file_path)
+        # there is a db index on fileloc_hash
+        rows = (
+            session.query(cls.dag_id)
+            .filter(cls.fileloc_hash == file_path_hash, cls.fileloc == dag_file_path)
+            .all()
+        )
+        dag_ids_from_db = [i[0] for i in rows]
```
Comment (ephraimbuddy) [inline airflow/models/serialized_dag.py]:
```suggestion
        dag_ids_from_db = [i.dag_id for i in rows]
```

File: airflow/models/serialized_dag.py [hunk 5]
```diff
@@ -220,6 +220,35 @@ def remove_dag(cls, dag_id: str, session: Session = None):
...
+
+        dag_ids_to_remove = [dag_id for dag_id in dag_ids_from_db if dag_id not in active_dag_ids]
+        if dag_ids_to_remove:
+            # lower the number of rows deleted at once
+            for dag_id in dag_ids_to_remove:
+                session.execute(cls.__table__.delete().where(cls.dag_id == dag_id))
```
Comment (ephraimbuddy) [inline airflow/models/serialized_dag.py:247]:
```suggestion
            session.query(cls.dag_id).filter(cls.dag_id.in_(dag_ids_to_remove)).delete(synchronize_session='fetch')
```

File: tests/dag_processing/test_processor.py [hunk 6]
```diff
@@ -635,6 +636,30 @@ def test_import_error_tracebacks(self, tmpdir):
             )
             session.rollback()
 
+    def test_process_file_should_remove_unknown_serialized_dags(self):
+        dag_file = os.path.join(
+            os.path.dirname(os.path.realpath(__file__)), '../dags/test_only_dummy_tasks.py'
+        )
```
Comment (ephraimbuddy) [inline tests/dag_processing/test_processor.py]:
```suggestion
        dag_file = os.path.join(TEST_DAGS_FOLDER, 'test_only_dummy_tasks.py')
```
Then import `TEST_DAGS_FOLDER` from `tests.models`

Comment (pingzh) [reply tests/dag_processing/test_processor.py]: nice, thanks

File: airflow/models/serialized_dag.py (hunk 4 above)
Comment (pingzh) [reply airflow/models/serialized_dag.py]: this is way better. thanks

**Review by KevinYang21 — APPROVED** [review]

**Comment by SamWheating** [comment]
Thoughts on deleting this DAG from the serialized DAGs table at the same time that it is marked inactive?

https://github.com/apache/airflow/blob/cfa95af7e83b067787d8d6596caa3bc97f4b25bd/airflow/dag_processing/manager.py#L508-L515

This query is gonna run really frequently (once for every DAG per parse interval) and most of the time won't have any missing DAGs to delete, by performing this on an intermittent / as-needed basis we can reduce the volume of queries on the database.

**Comment by pingzh** [comment]
> Thoughts on deleting missing DAGs from the serialized DAGs table at the same time that it is marked inactive?
> 
> https://github.com/apache/airflow/blob/cfa95af7e83b067787d8d6596caa3bc97f4b25bd/airflow/dag_processing/manager.py#L508-L515
> 
> There's some background in #21397 (Although I don't think the performance implications here are nearly as severe as in that case, as the serialized DAG table is properly indexed)
> 
> This query is gonna run really frequently (once for every DAG per parse interval) and most of the time won't have any missing DAGs to delete, by performing this on an intermittent basis rather than constantly, we can reduce the volume of queries.
> 
> This also would centralize similar pieces of code, rather than having similar functionality in completely separate parts of the codebase.

@SamWheating thanks for the context. that is a good idea to keep them together. updating the pr.

File: airflow/dag_processing/manager.py [hunk 7]
```diff
@@ -514,6 +514,10 @@ def _deactivate_stale_dags(self, session=None):
                 if deactivated:
                     self.log.info("Deactivated %i DAGs which are no longer present in file.", deactivated)
 
+                for dag_id in to_deactivate:
+                    SerializedDagModel.remove_dag(dag_id)
+                    self.log.info("Deleted DAG %s in serialized_dag table", dag_id)
+
```
Comment (kaxil) [inline airflow/dag_processing/manager.py:524]: Serialized DAGs for Deleted dag files are removed in the following place too

https://github.com/apache/airflow/blob/d8f8f8d3778fac8052862435e4e348f1f76f20ce/airflow/dag_processing/manager.py#L717-L722 

Does that not suffice?

Comment (pingzh) [reply airflow/dag_processing/manager.py:524]: @kaxil `remove_deleted_dags` only removes dags whose dag_files are deleted.

this pr takes care of the case like a dag_id is renamed in a dag_file, or some dag_ids are removed in a dag_file

Comment (kaxil) [reply airflow/dag_processing/manager.py:524]: Yup I know that :) -- What I more meant was do you think we should combine that or not run that now in Dag Processing Manager now that we have the changes in this PR? Don't see a point in having both the methods when one  in the PR is superset of the other.

File: airflow/dag_processing/manager.py [hunk 8]
```diff
@@ -514,6 +514,10 @@ def _deactivate_stale_dags(self, session=None):
                 if deactivated:
                     self.log.info("Deactivated %i DAGs which are no longer present in file.", deactivated)
 
+                for dag_id in to_deactivate:
```
Comment (kaxil) [inline airflow/dag_processing/manager.py:521]: Docstrings of the function should be updated to reflect this change too

Comment (pingzh) [reply airflow/dag_processing/manager.py:521]: good call

**Comment by KevinYang21** [comment]
Thanks everyone for help reviewing the PR. Given the PR has been open for a while and all comments have been addressed, if no more comments, I'll merge this PR tonight.

**Comment by potiuk** [comment]
> Thanks everyone for help reviewing the PR. Given the PR has been open for a while and all comments have been addressed, if no more comments, I'll merge this PR tonight.

Just released the providers RC (with some 'surprises') - I will take a look tomorrow finally if you can wait.

**Review by potiuk — APPROVED** [review]

**Pull request merged**